pkg/stream/environment_test.go

```diff
@@ -622,4 +622,124 @@ var _ = Describe("Environment", func() {
 			Eventually(logBuffer).Within(time.Millisecond * 500).Should(gbytes.Say(`"locator operation failed" error="err maybe later"`))
 		})
 	})
+
+	Context("query sequence", func() {
+		Context("input validation succeeds", func() {
+
+			BeforeEach(func() {
+				mockRawClient.EXPECT().
+					IsOpen().
+					Return(true) // from maybeInitializeLocator
+			})
+
+			It("queries last publishingid for a given producer and stream", func() {
+				// setup
+				publishingId := uint64(42)
+				mockRawClient.EXPECT().
+					QueryPublisherSequence(gomock.AssignableToTypeOf(ctxType), gomock.AssignableToTypeOf("string"), gomock.AssignableToTypeOf("string")).
+					Return(publishingId, nil)
+
+				// act
+				pubId, err := environment.QuerySequence(rootCtx, "producer-id", "stream-id")
+				Expect(err).ToNot(HaveOccurred())
+				Expect(pubId).To(BeNumerically("==", 42))
+
+			})
+
+			When("there is an error", func() {
+				It("bubbles up the error", func() {
+					// setup
+					publishingId := uint64(0)
+
+					mockRawClient.EXPECT().
+						QueryPublisherSequence(gomock.AssignableToTypeOf(ctxType), gomock.AssignableToTypeOf("string"), gomock.AssignableToTypeOf("string")).
+						Return(publishingId, errors.New("err not today")).
+						Times(3)
+
+					_, err := environment.QuerySequence(rootCtx, "producer-id", "stream-id")
+					Expect(err).To(MatchError("err not today"))
+				})
+
+			})
+
+			When("there are multiple locators", func() {
+				var (
+					locator2rawClient *stream.MockRawClient
+				)
+
+				BeforeEach(func() {
+					locator2rawClient = stream.NewMockRawClient(mockCtrl)
+					environment.AppendLocatorRawClient(locator2rawClient)
+					environment.SetBackoffPolicy(backOffPolicyFn)
+					environment.SetLocatorSelectSequential(true)
+
+					// have to set server version again because there's a new locator
+					// environment.SetServerVersion("3.11.1")
```

Comment:
Remove commented code, rebase v2 branch and ship it!

Reply:
Done and Done!